tests/port-forwarding-server.spec.ts

```diff
@@ -0,0 +1,133 @@
+      stdio: 'pipe',
+      detached: true,
+      env: {
+        ...process.env,
+        PW_SOCKS_PROXY_PORT: String(proxyPort)
+      }
+    });
+    this._driverProcess.unref();
+    this._receivedPortPromise = new Promise<string>((resolve, reject) => {
+      this._driverProcess.stdout.on('data', (data: Buffer) => {
+        const prefix = 'Listening on ';
+        const line = data.toString();
+        if (line.startsWith(prefix))
+          resolve(line.substr(prefix.length));
+      });
+      this._driverProcess.on('exit', () => reject());
+    });
+  }
+  async kill() {
+    const waitForExit = new Promise<void>(resolve =>  this._driverProcess.on('exit', () => resolve()));
+    this._driverProcess.kill('SIGKILL');
+    await waitForExit;
+  }
+  public async wsEndpoint(): Promise<string> {
+    return await this._receivedPortPromise;
+  }
+}
+
+const it = contextTest.extend<{ pageFactory: (redirectPortForTest?: number) => Promise<Page> }>({
+  pageFactory: async ({ browserName, browserOptions }, run, testInfo) => {
+    const browserServers: BrowserServer[] = [];
+    const playwrightServers: OutOfProcessPlaywrightServer[] = [];
+    await run(async (redirectPortForTest?: number): Promise<Page> => {
+      const server = new OutOfProcessPlaywrightServer(0, 3200 + testInfo.workerIndex);
+      playwrightServers.push(server);
+      const service = await PlaywrightClient.connect({
+        wsEndpoint: await server.wsEndpoint(),
+      });
+      const playwright = service.playwright();
+      playwright._enablePortForwarding(redirectPortForTest);
+      const browser = await playwright[browserName].launch(browserOptions);
+      return await browser.newPage();
+    });
+    for (const browserServer of browserServers)
+      await browserServer.close();
+    for (const playwrightServer of playwrightServers)
+      await playwrightServer.kill();
+  },
+});
+
+it.fixme(({ platform, browserName }) => platform === 'darwin' && browserName === 'webkit');
+it.skip(({ mode }) => mode !== 'default');
+
+async function startTestServer() {
+  const server = http.createServer((req: http.IncomingMessage, res: http.ServerResponse) => {
+    res.end('<html><body>from-retargeted-server</body></html>');
+  });
+  await new Promise(resolve => server.listen(0, resolve));
+  return {
+    testServerPort: (server.address() as net.AddressInfo).port,
+    stopTestServer: () => server.close()
+  };
+}
+
+it('should forward non-forwarded requests', async ({ pageFactory, server }) => {
+  let reachedOriginalTarget = false;
+  server.setRoute('/foo.html', async (req, res) => {
+    reachedOriginalTarget = true;
+    res.end('<html><body>original-target</body></html>');
+  });
+  const page = await pageFactory();
+  await page.goto(server.PREFIX + '/foo.html');
+  expect(await page.content()).toContain('original-target');
+  expect(reachedOriginalTarget).toBe(true);
```

Comment:
We should assert here that it used the proxy?

Reply:
Leaving it for a follow-up.